pkg/image/registry/imagestream/etcd/image.go

```diff
@@ -0,0 +1,224 @@
+	if i.informer == nil {
+		return nil, false, nil
+	}
+	return i.informer.GetStore().GetByKey(key)
+}
+func (i imageLayerIndex) Run(stopCh <-chan struct{}) {
+	if i.informer == nil {
+		return
+	}
+	i.informer.Run(stopCh)
+}
+
+// NewImageLayerIndex creates a new index over a store that must return
+// images.
+func NewImageLayerIndex(lw ImageListWatch) ImageLayerIndex {
+	informer := cache.NewSharedIndexInformer(&cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			list, err := lw.List(metav1.ListOptions{
+				ResourceVersion: options.ResourceVersion,
+				Limit:           options.Limit,
+				Continue:        options.Continue,
+			})
+			if err != nil {
+				return nil, err
+			}
+			// reduce the full image list to a smaller subset.
+			out := &metainternalversion.List{
+				Items: make([]runtime.Object, len(list.Items)),
+			}
+			out.Continue = list.Continue
+			out.ResourceVersion = list.ResourceVersion
+			for i, image := range list.Items {
+				out.Items[i] = &ImageLayers{
+					Name:     image.Name,
+					Layers:   image.DockerImageLayers,
+					Manifest: manifestFromImage(&image),
+				}
+			}
+			return out, nil
+		},
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			w, err := lw.Watch(metav1.ListOptions{
+				ResourceVersion: options.ResourceVersion,
+			})
+			if err != nil {
+				return nil, err
+			}
+			return watch.Filter(w, func(in watch.Event) (out watch.Event, keep bool) {
+				if in.Object == nil {
+					return in, true
+				}
+				// reduce each object to the minimal subset we need for the cache
+				image, ok := in.Object.(*imagev1.Image)
+				if !ok {
+					return in, true
+				}
+				layers := &ImageLayers{
+					Name:            image.Name,
+					ResourceVersion: image.ResourceVersion,
+					Layers:          image.DockerImageLayers,
+					Manifest:        manifestFromImage(image),
+				}
+				in.Object = layers
+				return in, true
+			}), nil
+		},
+	}, &ImageLayers{}, 0, cache.Indexers{
+		// layers allows fast access to the images with a given layer
+		"layers": func(obj interface{}) ([]string, error) {
+			entry, ok := obj.(*ImageLayers)
+			if !ok {
+				return nil, fmt.Errorf("unexpected cache object %T", obj)
+			}
+			keys := make([]string, 0, len(entry.Layers))
+			for _, layer := range entry.Layers {
+				keys = append(keys, layer.Name)
+			}
+			return keys, nil
+		},
+	})
+	return imageLayerIndex{informer: informer}
+}
+
+// manifestFromImage attempts to find a manifest blob description from
+// an image. Images older than schema2 in Docker do not have a manifest blob.
+func manifestFromImage(image *imagev1.Image) *imagev1.ImageLayer {
+	if image.DockerImageManifestMediaType != "application/vnd.docker.distribution.manifest.v2+json" {
+		return nil
+	}
+	meta := &imageapi.DockerImage{}
+	if _, _, err := legacyscheme.Codecs.UniversalDecoder().Decode(image.DockerImageMetadata.Raw, nil, meta); err != nil {
```

Comment:
You have the skills. You can make your own scheme for this package instead of relying on a global scheme.

Reply:
It was somewhat invasive (the docker api types needed to be split) so i moved it to #20241